Optimise batch-aware highly_variable_genes for Dask arrays

mumichae opened the pull request:
<!--
Thanks for opening a PR to scanpy!
Please be sure to follow the guidelines in our contribution guide (https://scanpy.readthedocs.io/en/latest/dev/index.html) to familiarize yourself with our workflow and speed up review.
-->

`sc.pp.highly_variable_genes` does not scale well when run per batch (`batch_key` is set) for Dask arrays. This PR refactors the per batch runs to get similar runtime as the non dask array version.

<!-- Please check (“- [x]”) and fill in the following boxes -->
- [x] Closes #3736
- [x] Tests included or not required because: no API changes
<!-- Only check the following box if you did not include release notes -->
- [ ] Release notes not necessary because:

mumichae committed:
simplify filtering check, make subsetting optional

mumichae committed:
refactor batch loop for more efficient handling of for dask arrays

mumichae committed:
dask import only when dask array is used

flying-sheep committed:
Merge branch 'main' into optimise_hvg_batch

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -437,14 +437,10 @@ def _highly_variable_genes_batched(
-    gene_list = adata.var_names
-    for batch in batches:
-        adata_subset = adata[adata.obs[batch_key] == batch]
+    def process_batch(batch_mask, adata, layer, gene_list, **hvg_kwargs):
```
Please add types. You can type `hvg_kwargs` by doing

```py
class HvgArgs(TypedDict):
    cutoff: _Cutoffs | int
    n_bins: int,
    flavor: flavor: Literal["seurat", "cell_ranger"]
```

and then use it in functions:

```py
def ...(..., **hvg_kwargs: Unpack[HvgArgs]) -> ...:
```

you could also update `_highly_variable_genes_batched` to use it that way.

ilan-gold replied:
Also probably don't want this inlined

mumichae replied:
I refactored that part of the code, but also cleaned up some dataframe handling, as well as making the filter call part of the `_highly_variable_genes_single_batch` directly, which could be considered new functionality. Lmk if things look OK or you want to preserve more of the old code

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -472,7 +469,33 @@ def _highly_variable_genes_batched(
+        from dask import compute, delayed
+
+        per_batch_func = delayed(process_batch)
+        _set_obs_rep(adata, X.persist(), layer=layer)
```
I still don’t know what `persist` is for, but @ilan-gold doesn’t like it I think. @ilan-gold, what are your objections again? Do they apply here?

ilan-gold replied:
It requires that you have enough memory to do this, hence my general opposition.  

@mumichae In your specific use-case, are you reading data off-disk or are you creating  dask arrays on top of in-memory arrays to get the memory benefit from concatenation (I see in `adata = ad.concat([adata_sub] * n_batches, label='batch', keys=range(n_batches))` you are doing things in memory).  I would ask then why this shouldn't be something the user does outside this function?

mumichae replied:
The in-memory example was just for playing around and figuring out the bottleneck. But I also observed the same issue with on-disk arrays, which is what I use in real-world workflows.

With persist, I guess it's a tradeoff of memory vs runtime, and for my examples things scaled accordingly (w/o persist has lower memory usage but takes longer and vice versa). But even without persist, the current performance is nowhere near as bad as before my fix, so I think it should be fine.

I removed the persist statement from the code and will just call persist in my workflows. I agree that it's better to let the user decide how they want the matrix stored.

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -472,7 +469,33 @@ def _highly_variable_genes_batched(
+        for batch in batches
+    ]
+    if not isinstance(dfs[0], pd.DataFrame):
+        dfs = compute(*dfs)
```
I think this is cleaner:

Do

```py
if is_dask := isinstance(X, DaskArray):
    from dask import delayed
```

above and then

```suggestion
    if is_dask:
        from dask import compute

        dfs = compute(*dfs)
```

maybe even do `dfs = (compute(df) for df in dfs)`?

if there are many, pandas might be able to use that more effectively, but maybe have dask schedule all the computes as it sees fit is better.

flying-sheep reviewed (commented):
This looks great! With a few tiny changes to the conditional import you mentioned, I’d say this is cleaner than before.

mumichae committed:
Update src/scanpy/preprocessing/_highly_variable_genes.py

Co-authored-by: Philipp A. <flying-sheep@web.de>

ilan-gold reviewed (commented):
Generally this looks great!

mumichae committed:
use generators for scheduling jobs with pandas.concat

mumichae committed:
do not persist dask arrays

mumichae committed:
create HvgArgs for managing arguments

mumichae committed:
refactor per batch hvg so that process_batch is replaced by _highly_variable_genes_single_batch; includes filtering unexpressed genes as part of _highly_variable_genes_single_batch

mumichae committed:
add release notes

flying-sheep committed:
fix typing

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -538,6 +573,7 @@ def highly_variable_genes(  # noqa: PLR0913
     subset: bool = False,
     inplace: bool = True,
     batch_key: str | None = None,
+    _filter_genes: bool = False,
```
this parameter is unused, what’s the plan here?

if this isn’t accidentally left in: if you want to introduce a new public parameter, it can’t have a name prefixed with `_`, if it’s for internal use only, it shouldn’t have documentation

mumichae replied:
I changed it to `filter_unexpressed_genes`. This parameter was introduced so that the gene filtering required by per-batch HVGs is possible. It also adds more functionality to the top-level `highly_variable_genes` function

flying-sheep replied:
I’m moving it into a new PR so this one can go into a patch

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -428,56 +467,52 @@ def _nth_highest(x: NDArray[np.float64] | DaskArray, n: int) -> float | DaskArra
+            adata=adata,
+            batch_mask=adata.obs[batch_key] == batch,
+            layer=layer,
+            gene_list=adata.var_names,
```
I don’t think this parameter exists!

mumichae replied:
Good point, I forgot to remove after refactoring.

```suggestion
```

flying-sheep committed:
always upload test results

flying-sheep committed:
also for min

mumichae committed:
Remove redundant parameter

mumichae committed:
rename _filtered_genes to filtered_unexpressed_genes

mumichae committed:
using walrus operator for type checking

mumichae committed:
Merge branch 'main' into optimise_hvg_batch

flying-sheep commented on src/scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -538,6 +572,7 @@ def highly_variable_genes(  # noqa: PLR0913
     subset: bool = False,
     inplace: bool = True,
     batch_key: str | None = None,
+    filter_unexpressed_genes: bool = False,
```
This parameter is still unused here, passing it to `highly_variable_genes` does nothing.

Can you please either remove it again or:

1. set the default to `None`, since it’s derived from `batch_key`.
2. explain in the parameter docs and a comment on the `else` branch where `_highly_variable_genes_batched` is called, that the parameter is always `True` if `batch_key` is `False`
3. raise an error if an incompatible combination is chosen (i.e. `batch_key="..."` and `filter_unexpressed_genes=False`) and add a test that this error is raised
4. actually pass it on to `_highly_variable_genes_single_batch` and add a test that checks that `highly_variable_genes(..., batch_key=None, filter_unexpressed_genes=True)` behaves differently from `highly_variable_genes(..., batch_key=None, filter_unexpressed_genes=False)`

mumichae replied:
Ah, I meant to pass it to the `_highly_variable_genes_single_batch` in the non-batch aware mode. I applied your suggestions, but used a warning instead of an error in case of incompatible configuration, as I don't think it warrants an error. But open to changing it to an error if that makes more sense from a usability standpoint

mumichae committed:
fix passing filter_unexpressed_genes to respective functions

mumichae committed:
fix typing for filter_unexpressed_genes, so that filter_unexpressed_genes=None gets interpreted correctly

mumichae commented:
@flying-sheep  I addressed your comments (thanks a lot for catching the issue btw, I completely missed it!)
But for some reasons I'm getting issues with the test upload. Any way you can rerun the github actions for the last commit? Or any input on what went wrong?

flying-sheep committed:
remove filter_unexpressed_genes for patch

flying-sheep committed:
Merge branch 'main' into optimise_hvg_batch

flying-sheep committed:
remove unused defaults
